### PR TITLE
[server] Introduce EntitlementServiceChargbee and move relevant parts of EligibilityService into it

### PR DESCRIPTION
**Pull request by geropl**
## Description
This PR does two things, while **not touching current behavior**:
 - it introduces the `EntitlementService` interface, which is based on a subset of methods of `EligibilityService`
 - it add a Chargebee implementation, which is 100% the identical code copied over from `EligibilityService`

## Related Issue(s)
Context: https://github.com/gitpod-io/gitpod/issues/11402

## How to test
 - open a workspace on this PR
 - get the content of the `EligiblityService` [before this commit](https://raw.githubusercontent.com/gitpod-io/gitpod/e3452c71b2fe1e81bfb91b5cc25bc8089d4efbde/components/server/ee/src/user/eligibility-service.ts), and put it into a file called `eligibilityservice.ts`
 - run `diff ee/src/billing/entitlement-service-chargebee.ts eligibilityservice.ts | less`
 - inspect the output carefully; notice how there is no relevant diff between the two

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

File: components/server/ee/src/billing/entitlement-service-chargebee.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+        return subscriptions.filter((s) => eligblePlans.includes(s.planId!)).length > 0;
+    }
+
+    /**
+     * Returns the default workspace timeout for the given user at a given point in time
+     * @param user
+     * @param date The date for which we want to know the default workspace timeout (depends on active subscription)
+     */
+    async getDefaultWorkspaceTimeout(user: User, date: Date = new Date()): Promise<WorkspaceTimeoutDuration> {
```
Comment (geropl) [inline components/server/ee/src/billing/entitlement-service-chargebee.ts:147]: This method received the only notable modification: It is a merge of `UserService.getDefaultWorkspaceTimeout` and `EligibilityService.getDefaultWorkspaceTimeout`

**Comment by geropl** [comment]
Noticed two small changes that I need to handle better. Will notify again once done. :+1:

File: components/server/ee/src/billing/entitlement-service-chargebee.ts (hunk 1 above)
Comment (geropl) [reply components/server/ee/src/billing/entitlement-service-chargebee.ts:160]: Ok, reverted that part. It's a plain copy now.

**Comment by geropl** [comment]
@AlexTugarev This is ready for review now! It should really not change any behavior the slightest, and is reduced to a mere move. :+1:

File: components/server/src/billing/entitlement-service.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    async mayStartWorkspace(
+        user: User,
+        date: Date,
+        runningInstances: Promise<WorkspaceInstance[]>,
+    ): Promise<MayStartWorkspaceResult> {
+        return { enoughCredits: true };
+    }
+
+    async maySetTimeout(user: User, date: Date): Promise<boolean> {
+        return true;
```
Comment (AlexTugarev) [inline components/server/src/billing/entitlement-service.ts:64]: really?

Comment (AlexTugarev) [reply components/server/src/billing/entitlement-service.ts:64]: I see. Existing strategy for disabled payment is to allow everything.

File: components/server/ee/src/user/eligibility-service.spec.db.ts [hunk 3]
```diff
@@ -147,26 +144,27 @@ class AccountServiceSpec {
...
-    async testUserGetsMoreResources_excludeFromMoreResources() {
-        await this.createTsSubscription(true);
-
-        const actual = await this.cut.userGetsMoreResources(this.user);
-        expect(
-            actual,
-            "user with Team Unleashed but excludeFromMoreResources set does not get 'more resources'",
-        ).to.equal(false);
-    }
+    // TODO(gpl) These should be moved over to EntitlementService.spec.ts
```
Comment (AlexTugarev) [inline components/server/ee/src/user/eligibility-service.spec.db.ts:147]: 🤞🏻

File: components/server/ee/src/user/user-service.ts [hunk 4]
```diff
@@ -72,9 +73,10 @@ export class UserServiceEE extends UserService {
         }
     }
 
+    // TODO(gpl) Needs to fold into EntitlementService
     async userGetsMoreResources(user: User): Promise<boolean> {
         if (this.config.enablePayment) {
```
Comment (AlexTugarev) [inline components/server/ee/src/user/user-service.ts:78]: `if (this.config.enablePayment)` still needed?

File: components/server/src/billing/entitlement-service.ts (hunk 2 above)
Comment (geropl) [reply components/server/src/billing/entitlement-service.ts:64]: Exactly. This implementation gathers defaults that so far have been hidden in various places.

:bulb: But notice that this is the code for the _community_ distribution which we do not even hand out. Self-Hosted is based on the EE code, which is driven by the "new" `EntintlementServiceChargebee` (the 100% old code).

In follow-up steps we want to separate that into pure `Chargebee` and `License`, but: step-by-step :footprints:

File: components/server/ee/src/user/user-service.ts [hunk 5]
```diff
@@ -17,24 +17,25 @@ import { inject } from "inversify";
...
     @inject(OssAllowListDB) protected readonly OssAllowListDb: OssAllowListDB;
     @inject(HostContextProvider) protected readonly hostContextProvider: HostContextProvider;
     @inject(Config) protected readonly config: Config;
 
+    // TODO(gpl) Needs to fold into EntitlementService
     async getDefaultWorkspaceTimeout(user: User, date: Date): Promise<WorkspaceTimeoutDuration> {
         if (this.config.enablePayment) {
             // the SaaS case
-            return this.eligibilityService.getDefaultWorkspaceTimeout(user, date);
+            return this.entitlementService.getDefaultWorkspaceTimeout(user, date);
```
Comment (AlexTugarev) [inline components/server/ee/src/user/user-service.ts:38]: `if (this.config.enablePayment)` still needed?

Comment (geropl) [reply components/server/ee/src/user/user-service.ts:38]: No, that is going to be handled by the BillingMode-guided implementation. But that's a follow-up PR (cmp. [comment](https://github.com/gitpod-io/gitpod/pull/11831/files/bd28cba0f9d730d84b730dc497832f900b245ef5#r937673793)).

For this PR it was more important to make the code 100% identical to the old impl in eligibility service.

File: components/server/ee/src/user/user-service.ts (hunk 4 above)
Comment (geropl) [reply components/server/ee/src/user/user-service.ts:78]: https://github.com/gitpod-io/gitpod/pull/11831/files/bd28cba0f9d730d84b730dc497832f900b245ef5#r937675184

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 6]
```diff
@@ -404,7 +406,7 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
      * gitpod.io Extension point for implementing eligibility checks. Throws a ResponseError if not eligible.
      */
     protected async maySetTimeout(user: User): Promise<boolean> {
-        return this.eligibilityService.maySetTimeout(user);
+        return this.entitlementService.maySetTimeout(user, new Date());
```
Comment (AlexTugarev) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:409]: looking at `maySetTimeout`, the date param seems to be weird. it's used internally to express "until now", thus `new Date()`. but it doesn't seem to make any sense for `maySetTimeout`, does it? e.g. `maySetTimeout(user, yesterday)` is irrelevant. let's remove it, wdyt?

**Review by AlexTugarev — APPROVED** [review]
Awesome! Thanks for the first chunk!

/hold 
just in case you want to do something about https://github.com/gitpod-io/gitpod/pull/11831/files#r937678415

File: components/server/ee/src/workspace/gitpod-server-impl.ts (hunk 6 above)
Comment (geropl) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:409]: > but it doesn't seem to make any sense for maySetTimeout, does it? 

:information_source:  It's mainly there for testability.

But I agree, it should have a default to not bother clients. Will change in a follow-up! :+1:

**Comment by geropl** [comment]
Thank you for your review @AlexTugarev ! :pray: 

/unhold
to move forward and triple-check on staging.

**Pull request merged**